src/main/java/com/louis/HashTable.java

```diff
@@ -0,0 +1,272 @@
+	 * @throws IllegalArgumentException if key is null.
+	 */
+	public boolean contains(String key) {
+		Preconditions.checkArgument(key != null,
+				"First argument to contains(String key) is null");
+		int index = calculateDesiredPositionOfKey(key, currentCapacity);
+		return searchForKeyFromIndex(key, index).isPresent();
+	}
+
+	private int calculateDesiredPositionOfKey(String key, int modulus) {
+		return hash(key) % modulus;
+	}
+
+	private Optional<Integer> searchForKeyFromIndex(String key, int index) {
+        if (table[index] == null) {
+            return Optional.empty();
+        }
+		if (key.equals(table[index].getKey())) {
+			return Optional.of(index);
+		} else {
+			return searchForKeyForwardOfIndex(key, ((index + 1) % currentCapacity), index);
+		}
+	}
+
+	private Optional<Integer> searchForKeyForwardOfIndex(String key, int currentIndex,
+															 int finalIndex) {
+        if (table[currentIndex] == null) {
+            return Optional.empty();
+        }
+		if (key.equals(table[currentIndex].getKey())) {
+			return Optional.of(currentIndex);
+		} else {
+			if (currentIndex == finalIndex) {
+				return Optional.empty();
+			} else {
+				return searchForKeyForwardOfIndex(key, ((currentIndex + 1) % currentCapacity), finalIndex);
+			}
+		}
+	}
+
+	/**
+	 * Returns the value associated with a key.
+	 *
+	 * @param key The key for which to look up the value.
+	 * @return The value associated with the key or null if no such value.
+	 * @throws IllegalArgumentException if key is null.
+	 */
+	public synchronized String get(String key) {
+		Preconditions.checkArgument(key != null, "first argument to get(String key) is null");
+		int startIndex = calculateDesiredPositionOfKey(key, currentCapacity);
+		Optional<Integer> index = searchForKeyFromIndex(key, startIndex);
+		return index.isPresent() ? table[index.get()].getValue() : null;
+	}
+
+	/**
+	 * Adds a key and its value to the table.
+	 *
+	 * @param key is the key to be added.
+	 * @param value is the value for this key.
+	 * @throws IllegalArgumentException if key or value is null.
+	 */
+	public synchronized void put(String key, String value) {
+		Preconditions.checkArgument(key != null, "first argument to put(String key, ...) is null");
+		int startIndex = calculateDesiredPositionOfKey(key, currentCapacity);
+		Preconditions.checkArgument(value != null, "second argument to put(..., String value) is null");
+		Optional<Integer> index = putKeyFromIndex(key, startIndex, table, currentCapacity);
+
+        if (!index.isPresent()) {
+            resizeIfRequired();
+            put(key, value);
+            return;
+        }
+
+		// Increases the size if a new key is being put in.
+		if (table[index.get()] == null) {
+			currentSize++;
+		}
+		table[index.get()] = ImmutableHashEntry.builder().key(key).value(value).build();
+
+		resizeIfRequired();
+	}
+
+	private Optional<Integer> putKeyFromIndex(String key, int index, HashEntry[] entries, int capacity) {
```

Comment:
Adjust this to return an Integer rather than an optional - you should never fail to insert an object unless you run out of memory in the JVM, in which case the process will exit.